pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,302 @@
+		leaderworkerset.RevisionKey: revisionKey,
+	}})
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := ListRevisions(ctx, k8sClient, lws, selector)
+	if err != nil {
+		log.Error(err, "Listing all controller revisions")
+		return nil, err
+	}
+
+	if len(revisions) == 0 {
+		return nil, nil
+	}
+
+	if len(revisions) > 1 {
+		// Since we only create a controllerRevision when the template hash changes, only one should match
+		log.Error(err, "More than one revision exists for the given template hash; returning the latest revision")
+		return revisions[len(revisions)-1], nil
+	}
+
+	return revisions[0], nil
+}
+
+func GetRevisionKey(obj metav1.Object) string {
+	if obj.GetLabels() != nil {
+		return obj.GetLabels()[leaderworkerset.RevisionKey]
+	}
+	return ""
+}
+
+func CreateRevision(
+	ctx context.Context,
+	k8sClient client.Client,
+	revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
+	if err := k8sClient.Create(ctx, revision); err != nil {
+		return nil, err
+	}
+	created := &appsv1.ControllerRevision{}
+	if err := k8sClient.Get(ctx, types.NamespacedName{Namespace: revision.Namespace, Name: revision.Name}, created); err != nil {
+		return nil, err
+	}
+	return created, nil
+}
+
+// NewRevision instantiates a new ControllerRevision containing a patch that reapplies the target state of LeaderWorkerSet.
+// The Revision of the returned ControllerRevision is set to revision. If the returned error is nil, the returned
+// ControllerRevision is valid. LeaderWorkerSet revisions are stored as patches that re-apply the current state of set
+// to a new LeaderWorkerSet using a strategic merge patch to replace the saved state of the new LeaderWorkerSet.
+func NewRevision(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, revisionKey string) (*appsv1.ControllerRevision, error) {
+	var controllerKind = leaderworkerset.GroupVersion.WithKind("LeaderWorkerSet")
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	}})
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := ListRevisions(ctx, k8sClient, lws, selector)
+	revision := nextRevision(revisions)
+	if err != nil {
+		return nil, err
+	}
+	patch, err := getPatch(lws)
+	if err != nil {
+		return nil, err
+	}
+
+	templateLabels := map[string]string{
+		leaderworkerset.RevisionKey:     revisionKey,
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	}
+
+	cr := &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:          templateLabels,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(lws, controllerKind)},
+			Namespace:       lws.Namespace,
+		},
+		Data:     runtime.RawExtension{Raw: patch},
+		Revision: revision,
+	}
+
+	hash := hashRevision(cr)
+	cr.Name = revisionName(lws.Name, hash, revision)
+	if cr.Labels[leaderworkerset.RevisionKey] == "" {
+		cr.Labels[leaderworkerset.RevisionKey] = hash
+	}
```

Comment:
```suggestion
	hash := hashRevision(cr)	
	if revisionKey == "" {
		revisionKey = hash
	}
	cr := &appsv1.ControllerRevision{
		ObjectMeta: metav1.ObjectMeta{
			Name: revisionName(lws.Name, hash, revision),
			Namespace:       lws.Namespace,
			Labels:          map[string]string{
		            leaderworkerset.RevisionKey:     revisionKey,
		            leaderworkerset.SetNameLabelKey: lws.Name,
	                 },
			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(lws, controllerKind)},
		},
		Data:     runtime.RawExtension{Raw: patch},
		Revision: revision,
	}

```

Reply:
We need to define the variable `cr` before we can calculate the hash. At line 143, `cr` still hasn't been defined yet.

Reply:
ah, right; perhaps we can still cleanup the label setting a bit

```
cr := &appsv1.ControllerRevision{
		ObjectMeta: metav1.ObjectMeta{
			Namespace:       lws.Namespace,
			Labels:          map[string]string{
		            leaderworkerset.SetNameLabelKey: lws.Name,
	                 },
			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(lws, controllerKind)},
		},
		Data:     runtime.RawExtension{Raw: patch},
		Revision: revision,
	}
	hash := hashRevision(cr)	
	if revisionKey == "" {
		revisionKey = hash
	}
        cr.Name = revisionName(lws.Name, hash, revision)
        cr.Labels[leaderworkerset.RevisionKey] = revisionKey
```